pkg/alertmanager/multitenant.go

```diff
@@ -1089,6 +1186,39 @@ func (am *MultitenantAlertmanager) getPerUserDirectories() map[string]string {
 	return result
 }
 
+// UpdateState implements the Alertmanager service.
+func (am *MultitenantAlertmanager) ReadState(ctx context.Context, req *alertmanagerpb.ReadStateRequest) (*alertmanagerpb.ReadStateResponse, error) {
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("no user id found in context")
+	}
+
+	am.alertmanagersMtx.Lock()
+	defer am.alertmanagersMtx.Unlock()
```

Comment:
Do we need to block until the end of the call? I think you just need to protect the access to `am.alertmanagers[userID]`, so you should be able to unlock the line after that. Am I missing anything?

Reply:
Have to admit I have not thought about it  - I was following on from the `UpdateState` implementation. I'll take a closer look to see if the mutex is protecting us from anything other than the map access.